test/integration/util/cloud.go

```diff
@@ -0,0 +1,86 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"time"
+
+	"github.com/golang/glog"
+
+	"golang.org/x/oauth2"
+	"k8s.io/kubernetes/pkg/cloudprovider/providers/gce"
+)
+
+const (
+	// TestProjectID is the project id used for creating NewMockGCECloud
+	TestProjectID = "test-project"
+	// TestNetworkProjectID is the network project id for creating NewMockGCECloud
+	TestNetworkProjectID = "net-test-project"
+	// TestRegion is the region for creating NewMockGCECloud
+	TestRegion = "test-region"
+	// TestZone is the zone for creating NewMockGCECloud
+	TestZone = "test-zone"
+	// TestNetworkName is the network name for creating NewMockGCECloud
+	TestNetworkName = "test-network"
+	// TestSubnetworkName is the sub network name for creating NewMockGCECloud
+	TestSubnetworkName = "test-sub-network"
+	// TestSecondaryRangeName is the secondary range name for creating NewMockGCECloud
+	TestSecondaryRangeName = "test-secondary-range"
+)
+
+type mockTokenSource struct{}
+
+func (*mockTokenSource) Token() (*oauth2.Token, error) {
+	return &oauth2.Token{
+		AccessToken:  "access",
+		TokenType:    "Bearer",
+		RefreshToken: "refresh",
+		Expiry:       time.Now().Add(1 * time.Hour),
+	}, nil
+}
+
+// NewMockGCECloud returns a handle to a GCECloud instance that is
+// served by a mock http server
+func NewMockGCECloud(handler http.Handler) (ShutdownFunc, *gce.GCECloud, error) {
```

Comment:
Have you taken a look at:
- https://github.com/kubernetes/kubernetes/tree/master/pkg/cloudprovider/providers/fake
- https://github.com/kubernetes/kubernetes/tree/master/pkg/cloudprovider/providers/gce/cloud/mock

Maybe we can just extend sth there?

Reply:
From what I could see, the fake implementations were mocking the client. The intent here was to mock the server and run the whole of the gce cloud client code. For example, some calls to GCE end up being multiple calls to the cloud under the hoods and part of the idea was to be able to measure that. The actual measurement is TBD enhancement, but the infra here allows for that.

Reply:
SG.

Reply:
Actually, I'm not convinced mocking the server is the right thing to do here. Your mock server will return immediately with the HTTP result anyway and if you need to tune the delay on the GCE operation, you could insert a delay on the client-side.